### PR TITLE
Update evaluate function & metrics.py

### PR DESCRIPTION
**Pull request by Eleven1Liu**
1. Remove unused function in `MultiLabelMetrics`.
2. `metric.py`: `len(tk)` to k
3. Output prediction (provide top1 ~top 100 with scores) in this format: 
`<label1>:<value1> <label2>:<value2> ...`

### CODE REVIEW

File: metrics.py [hunk 1]
```diff
@@ -30,13 +21,12 @@ def precision_at_k(y_true, y_pred_vals, k):
     # get precision at k for each example
     vals = []
     for i, tk in enumerate(topk):
-        if len(tk) > 0:
-            num_true_in_top_k = y_true[i,tk].sum()
-            denom = len(tk)
-            vals.append(num_true_in_top_k / float(denom))
+        num_true_in_top_k = y_true[i,tk].sum()
+        vals.append(num_true_in_top_k / float(k))
```
Comment (sian-chen) [inline metrics.py:25]: We don't need float() here, as k is already an integer.
```suggestion
        vals.append(num_true_in_top_k / k)
```

File: utils/utils.py [hunk 2]
```diff
@@ -71,3 +73,22 @@ def dump_log(config, metrics, split):
...
+
+    if config.predict_path:
+        predict_path = config.predict_path
+    else:
+        predict_path = os.path.join(config.result_dir, config.run_name, 'prediction.txt')
+    
+    os.makedirs(os.path.dirname(predict_path), exist_ok=True)
+    print(f'Dump top {k} prediction to {predict_path}.')
+    with open(predict_path, 'w') as fp:
+        for pred in np.vstack(y_pred):
```
Comment (sian-chen) [inline utils/utils.py:91]: Why we need vstack here?

Comment (Eleven1Liu) [reply utils/utils.py:91]: Oh! That a long story ... We do `add_batch` in `MultiLabelMetrics`. I will check this again.

File: metrics.py [hunk 3]
```diff
@@ -2,26 +2,17 @@
...
+from sklearn.metrics import precision_recall_fscore_support
 
-def macro_recall(y_true, y_pred):
-    recall_per_label = intersection(y_true, y_pred, 0) / (y_true.sum(axis=0) + 1e-10)
-    return np.mean(recall_per_label)
 
 def macro_f1(y_true, y_pred):
-    prec = macro_precision(y_true, y_pred)
-    rec = macro_recall(y_true, y_pred)
+    prec, rec, _, _ = precision_recall_fscore_support(y_true, y_pred, average='macro')
```
Comment (cjlin1) [inline metrics.py:11]: I now think we can do macro_prec, macro_rec to clearly see this is the average of prec or rec across labels

File: metrics.py [hunk 4]
```diff
@@ -2,26 +2,17 @@
...
     f1 = 2 * (prec * rec) / (prec + rec + 1e-10)
     return f1
 
+
```
Comment (cjlin1) [inline metrics.py:15]: why this empty line?

File: metrics.py [hunk 5]
```diff
@@ -30,13 +21,12 @@ def precision_at_k(y_true, y_pred_vals, k):
...
-        if len(tk) > 0:
-            num_true_in_top_k = y_true[i,tk].sum()
-            denom = len(tk)
-            vals.append(num_true_in_top_k / float(denom))
+        num_true_in_top_k = y_true[i,tk].sum()
+        vals.append(num_true_in_top_k / k)
 
     return np.mean(vals)
 
+
```
Comment (cjlin1) [inline metrics.py:29]: why this empty line?

File: utils/utils.py [hunk 6]
```diff
@@ -71,3 +73,22 @@ def dump_log(config, metrics, split):
         result[split] = [metrics]
     with open(log_path, 'w') as fp:
         json.dump(result, fp)
+
+
+def dump_top_k_prediction(config, classes, y_pred, k=100):
+    """Dump top k prediction to the predict_path. The format of this file is:
```
Comment (cjlin1) [inline utils/utils.py:79]:
```suggestion
    """Dump top k predictions to the predict_out_path. The format of this file is:
```

File: utils/utils.py [hunk 7]
```diff
@@ -71,3 +73,22 @@ def dump_log(config, metrics, split):
...
+    <label1>:<value1> <label2>:<value2> ...
+    """
+
+    if config.predict_path:
+        predict_path = config.predict_path
+    else:
+        predict_path = os.path.join(config.result_dir, config.run_name, 'prediction.txt')
```
Comment (cjlin1) [inline utils/utils.py:86]: predictions.txt

File: utils/utils.py [hunk 8]
```diff
@@ -71,3 +73,22 @@ def dump_log(config, metrics, split):
...
+    
+    os.makedirs(os.path.dirname(predict_path), exist_ok=True)
+    print(f'Dump top {k} prediction to {predict_path}.')
```
Comment (cjlin1) [inline utils/utils.py:89]: predictions

**Review by cjlin1 — APPROVED** [review]

File: metrics.py (hunk 4 above)
Comment (Eleven1Liu) [reply metrics.py:15]: two empty lines between functions
if functions are in a class, one empty line

It's a coding style we used before, we could discuss about this.

File: evaluate.py [hunk 9]
```diff
@@ -69,6 +69,9 @@ def eval(self, y_true, y_pred, threshold=0.5):
 
         return result
 
+    def get_y_pred(self):
+        return np.vstack(self.y_pred)
```
Comment (sian-chen) [inline evaluate.py:73]: Why do we need vstack here?

File: utils/utils.py [hunk 10]
```diff
@@ -78,6 +78,11 @@ def dump_log(config, metrics, split):
 def dump_top_k_prediction(config, classes, y_pred, k=100):
     """Dump top k predictions to the predict_out_path. The format of this file is:
     <label1>:<value1> <label2>:<value2> ...
+
+    Parameters:
+    classes (list): list of class names
+    y_pred (list): list of ndarray (shape: number of samples * number of classes)
```
Comment (sian-chen) [inline utils/utils.py:84]: The input is not a list of ndarray if we use vstack.
But do we really need vstack?

Comment (Eleven1Liu) [reply utils/utils.py]: As discussed, I've added more commands on these functions. Please check for updates.

**Review by sian-chen — APPROVED** [review]
